deployer.go

```diff
@@ -63,253 +67,282 @@ func newDeployer() (*deployer, error) {
 }
 
 func (d *deployer) deployAll() error {
-	// Get service definition - wanted units
-	//get whitelist
-	wantedUnits, zddUnits, err := d.buildWantedUnits()
-	deployedUnits := make(map[string]bool)
-
+	wantedServiceGroups, err := d.buildWantedUnits()
 	if err != nil {
 		return err
 	}
 
-	// create any missing units
-	for _, u := range wantedUnits {
-		//basically reproduce the behavior of the deployUnit() function but with the
-		//sequential deployment thrown in
+	toDelete := d.identifyDeletedServiceGroups(wantedServiceGroups)
+	toCreate := d.identifyNewServiceGroups(wantedServiceGroups)
+	purgeProcessed(wantedServiceGroups, toCreate)
+	toUpdateRegular, toUpdateSKRegular, toUpdateSequentially, toUpdateSKSequentially := d.identifyUpdatedServiceGroups(wantedServiceGroups)
 
-		isNew, err := d.isNewUnit(u)
-		if err != nil {
-			log.Printf("WARNING Failed to determine if it's a new unit %s: %v [SKIPPING]", u.Name, err)
-			continue
+	d.createServiceGroups(toCreate)
+	d.updateServiceGroupsNormally(toUpdateRegular)
+	d.updateServiceGroupsSequentially(toUpdateSequentially)
+	d.updateServiceGroupsSKsOnly(toUpdateSKRegular)
+	d.updateServiceGroupsSKsSequentially(toUpdateSKSequentially)
+	d.deleteServiceGroups(toDelete)
+
+	toLaunch := mergeMaps(toCreate, toUpdateRegular, toUpdateSKRegular)
+	err = d.launchAll(toLaunch)
```

Comment:
Would it make sense to move the launch logic in createServiceGroups, updateServiceGroupsNormally, updateServiceGroupsSKsOnly? This way you could easily run them concurrently.